### ISSUE TITLE
distance_transform returns NaN gradients wherever its convolution hits zero (log(0) through nan_to_num)

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4183]
## Summary

`kornia.contrib.distance_transform` returns **NaN** gradients for any input that has a region far enough from a boundary pixel for the cascade's convolution to return exactly zero — which includes the ordinary case of a sparse binary mask, and every all-zero input. The forward result is correct; only the backward is poisoned, silently, with no warning.

## Reproduction

`main` @ `ee6b713f`, torch 2.9.1, CPU:

```python
import torch
from kornia.contrib import distance_transform

torch.manual_seed(0)
x = (torch.rand(1, 1, 16, 16) > 0.9).float().requires_grad_()
distance_transform(x, kernel_size=3).sum().backward()
print(torch.isnan(x.grad).sum().item(), "of", x.grad.numel())   # 193 of 256
```

| input (float64) | NaN entries in the gradient |
|---|---|
| random binary mask, 16×16, ~10% ones | 175 / 256 |
| one-hot 4×4 (single boundary pixel) | 15 / 16 |
| all-zeros 4×4 | 16 / 16 |

float32 on the same random mask: 193 / 256. All NaN, no Inf.

## Mechanism

`kornia/contrib/distance_transform.py`, both the 2D and the 3D impl:

```python
cdt = filter2d(boundary, kernel, border_type="replicate")
cdt = -h * torch.log(cdt)
cdt = torch.nan_to_num(cdt, nan=0.0, posinf=0.0, neginf=0.0)
```

Where the convolution returns exactly `0`, `log` gives `-inf`, and `nan_to_num` cleans the *forward* value to `0`. Its backward, however, multiplies the incoming gradient by `isfinite(input)`, so the `inf` derivative of `log` at zero meets a zero multiplier and yields `0 * inf = NaN`. Isolated:

```python
x = torch.zeros(1, 1, 4, 4, dtype=torch.float64, requires_grad=True)
cdt = filter2d(x, torch.ones(1, 3, 3, dtype=torch.float64), border_type="replicate")
torch.nan_to_num(-0.35 * torch.log(cdt), nan=0.0, posinf=0.0, neginf=0.0).sum().backward()
print(x.grad.flatten()[:3])   # tensor([nan, nan, nan], dtype=torch.float64)
```

## Why the existing tests do not catch it

`tests/contrib/test_conv_distance_transformer.py::test_gradcheck` uses `torch.ones(B, C, H, W)` for both the 2D and the 3D case. With an all-ones image the convolution is positive everywhere, the `log` never sees zero, and the defective branch is never entered. The regression test for this needs an input with an actual empty region — a sparse mask or a one-hot image.

## Candidate fix

Guarding the `log` instead of cleaning up after it removes every NaN and leaves the forward **byte-identical**. Measured on the three inputs in the table above, against the library's own output (`torch.equal`, float64):

```python
positive = cdt > 0
cdt = -h * torch.log(torch.where(positive, cdt, torch.ones_like(cdt)))
cdt = torch.where(positive, cdt, torch.zeros_like(cdt))
```

| input | forward byte-identical to current | NaN grads now | NaN grads with the guard |
|---|---|---|---|
| random mask | yes | 193 | **0** |
| one-hot | yes | 15 | **0** |
| all zeros | yes | 16 | **0** |

Note `mask = cdt > 0` is computed *after* this line, on the cleaned value, so a naive `cdt.clamp_min(eps)` before the `log` is **not** a valid alternative: it turns those positions into `-h*log(eps) > 0` and flips them into the mask, changing the forward. Whatever the fix, the forward has to be shown byte-identical.

## Classification

**Broken output** (gradients), reachable from ordinary usage of a differentiable op. Not related to #4152 — verified present and identical on both sides of #4158, which changes nothing here.

## Environment

kornia @ `ee6b713f`, torch 2.9.1, Python 3.11.14, macOS/arm64, CPU.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by Jeevang1-epic** [comment]
@ducha-aiki  I'd like to take this one as well if it's still avalable.I went through the reproduction and the failure mechanism. Theforward path is being cleaned by `nan_to_num`, but when the convolution reaches exactly zero the backward still passes through `log(0)`, so the non-finite derivative ends up producing NaN gradients. I also see why the current gradcheck misses it: the all-ones input never exercises the zero-convolution branch.
My plan would be to first pin the regression with sparse/one-hot inputs, then update the 2D and 3D paths so the logarithm is only evaluated where `cdt > 0`. I'll verify that the forward output stays byte-identical to the current implementation, that the gradients are finite for the affected cases, and that the change remains compatible with the separate underflow work in #4158.If that scope looks right, could you claim this one for me too?

**Comment by ducha-aiki** [comment]
@Jeevang1-epic yes, take it — it is yours. GitHub will not let me put a non-collaborator in the assignee field, so this comment is the claim marker. You turned #4058 around in a day with a pin that actually fails on the merge-base, so I am happy to hand you the harder one.

Your reading is right, including the part most people miss: `nan_to_num` cleans the **forward** and the backward still multiplies by `isfinite(input)`, so `0 * inf = NaN` survives it. And you are right about why `test_gradcheck` cannot see it — `torch.ones(...)` never lets the convolution reach zero.

The scope you propose is the scope. Three things to hold on to while you do it:

1. **Do not `clamp_min` the convolution result before the `log`.** `mask = cdt > 0` is computed *after* the cleanup, so clamping to `eps` flips those pixels into the mask and changes the forward. The issue's candidate guard evaluates the `log` only where `cdt > 0` for exactly that reason; if you find a better shape, the test is the same — the forward must not move.
2. **Byte-identical, not close.** Compare `torch.equal` (or an int-view comparison) against a merge-base worktree on the three probes in the issue — a random binary mask, a one-hot, and an all-zeros input — not `assert_close`. The numbers I recorded there (193/256, 15/16, 16/16 NaN grads) are the fail-on-base evidence your pin has to reproduce.
3. **#4158 is a different defect in the same function.** It is approved and waiting on a CHANGELOG-only rebase, so assume it lands first and rebase onto it. Its fix is an upcast in `_check_h_range` / the kernel construction; yours is in the `log` backward. They should not collide, but say so explicitly in the PR once you have checked, rather than leaving it to the reviewer.

One extra that would be welcome but is not required: `test_gradcheck` feeding `torch.ones` is what let this hide for so long. If your new pin replaces that input rather than sitting beside it, say so — that is a coverage improvement in its own right.

Ping me when it is up.

Posted on behalf of @ducha-aiki by Claude (Opus 5).
